### PR TITLE
[mkldnn-v1.0] Add MKL-DNN Pooling

### PR DESCRIPTION
**Pull request by rongzha1**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here
@ciyongch @ZhennanQin @TaoLv @wuxun-zhang @PatricZhao

### CODE REVIEW

**Comment by pengzhao-intel** [comment]
@TaoLv take a review?

File: src/operator/nn/mkldnn/mkldnn_pooling.cc [hunk 1]
```diff
@@ -333,27 +292,31 @@ MKLDNNPoolingBwd &GetPoolingBwd(const PoolingParam &param,
 
   auto it = pooling_bwds.find(key);
   if (it == pooling_bwds.end()) {
-    auto diff_dst_mem = out_grad.GetMKLDNNData();
+    // mkldnn v1.0 add reoder to workaround testcase:test_make_subgraph;
+    // alread fixed in v1.1, will remove after v1.1 is integrated.
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_pooling.cc]: Please remove the comments.

File: src/operator/nn/mkldnn/mkldnn_pooling.cc [hunk 2]
```diff
@@ -401,14 +363,21 @@ void MKLDNNPoolingGradCompute(const OpContext &ctx, const PoolingParam &param,
...
-  MKLDNNStream::Get()->RegisterPrim(bwd.GetBwd());
+  MKLDNNStream::Get()->RegisterPrimArgs(bwd.GetBwd(), args);
   CommitOutput(in_grad, diff_src_mem);
   MKLDNNStream::Get()->Submit();
 }
 
 }  // namespace op
 }  // namespace mxnet
-#endif  // MXNET_USE_MKLDNN == 1
+#endif  // MXNET_USE_MKLDNN == 0
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_pooling.cc]: Why?

File: src/operator/nn/mkldnn/mkldnn_pooling.cc [hunk 3]
```diff
@@ -401,14 +363,21 @@ void MKLDNNPoolingGradCompute(const OpContext &ctx, const PoolingParam &param,
...
   auto &bwd = GetPoolingBwd(param, in_data, in_grad, out_grad);
   auto diff_src_mem =
-      CreateMKLDNNMem(in_grad, bwd.pd.diff_src_primitive_desc(), req);
+      CreateMKLDNNMem(in_grad, bwd.pd.diff_src_desc(), req);
+
+  mkldnn_args_map_t args = {
+    {MKLDNN_ARG_DIFF_DST, *(out_grad.GetMKLDNNData())},
+    {MKLDNN_ARG_DIFF_SRC, *diff_src_mem.second },
+  };
+  if (workspace != nullptr) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_pooling.cc]: Also check `with_workspace`?

File: src/operator/nn/mkldnn/mkldnn_pooling.cc (hunk 1 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_pooling.cc]: done

File: src/operator/nn/mkldnn/mkldnn_pooling.cc (hunk 2 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_pooling.cc]: typo

File: src/operator/nn/mkldnn/mkldnn_pooling.cc (hunk 3 above)
Comment (rongzha1) [reply src/operator/nn/mkldnn/mkldnn_pooling.cc]: OK. Done

**Comment by TaoLv** [comment]
Merge for now as the onnx test error is irrelevant and tracked by https://github.com/apache/incubator-mxnet/issues/16394.

**Pull request merged**